concordium-consensus/tests/scheduler/SchedulerTests/Delegation.hs

```diff
@@ -527,6 +572,348 @@ testCase9 _ pvString =
+        BS.PersistentBlockState pv ->
+        Helpers.PersistentBSM pv Assertion
+    checkState result blockState =
+        Helpers.assertBlockStateInvariantsH blockState (Helpers.srExecutionCosts result)
+    events =
+        [ BakerRemoved 4 baker4Address,
+          DelegationAdded 4 baker4Address,
+          DelegationSetDelegationTarget 4 baker4Address (DelegateToBaker 2),
+          DelegationSetRestakeEarnings 4 baker4Address False,
+          DelegationStakeIncreased 4 baker4Address 1_000
+        ]
+
+-- | Add delegator with 0 stake when already a baker should get rejected with
+--  `AlreadyABaker` in protocols <= P6 and `InsufficientDelegationStake` from P7.
+testCase11A ::
+    forall pv.
+    (IsProtocolVersion pv, PVSupportsDelegation pv) =>
+    SProtocolVersion pv ->
+    String ->
+    Spec
+testCase11A spv pvString =
+    specify (pvString ++ ": Add delegator with 0 stake when already baker should get rejected.") $ do
+        let transactions =
+                [ Runner.TJSON
+                    { payload =
+                        Runner.ConfigureDelegation
+                            { cdCapital = Just 0,
+                              cdRestakeEarnings = Just False,
+                              cdDelegationTarget = Just (DelegateToBaker 2)
+                            },
+                      metadata = makeDummyHeader baker4Address 1 1_000,
+                      keys = [(0, [(0, baker4KP)])]
+                    }
+                ]
+        (result, doBlockStateAssertions) <-
+            Helpers.runSchedulerTestTransactionJson
+                Helpers.defaultTestConfig
+                (initialBlockState2 @pv)
+                (Helpers.checkReloadCheck checkState)
+                transactions
+        let reason = case sSupportsFlexibleCooldown (sAccountVersionFor spv) of
+                SFalse -> AlreadyABaker 4
+                STrue -> InsufficientDelegationStake
+        Helpers.assertRejectWithReason reason result
+        doBlockStateAssertions
+  where
+    checkState ::
+        Helpers.SchedulerResult ->
+        BS.PersistentBlockState pv ->
+        Helpers.PersistentBSM pv Assertion
+    checkState result blockState = do
+        invariants <- Helpers.assertBlockStateInvariantsH blockState (Helpers.srExecutionCosts result)
+        updatedBaker4 <- BS.toTransientAccount . fromJust =<< BS.bsoGetAccountByIndex blockState 4
+        initialBaker4 <- BS.toTransientAccount =<< baker4Account @(AccountVersionFor pv)
+        return $ do
+            invariants
+            assertEqual
+                "Baker account should not have changed except nonce and balance"
+                ( initialBaker4
+                    & Transient.accountNonce .~ 2
+                    & Transient.accountAmount -~ Helpers.energyToAmount (Cost.configureDelegationCost + Cost.baseCost 81 1)
+                )
+                updatedBaker4
+
+-- | Reduce stake while in cooldown.
+testCase12 ::
```

Comment:
Can we have more descriptive names for these tests?

Reply:
Good idea. Added descriptive names.